lighthouse-core/gather/gatherers/image-elements-snapshot.js

```diff
@@ -0,0 +1,376 @@
+      } catch (_) {
+        // determineNaturalSize fails on invalid images, which we treat as non-visible
+      }
+    }
+
+    if (!size) return;
+    element.naturalDimensions = {width: size.naturalWidth, height: size.naturalHeight};
+  }
+
+  /**
+   * Images might be sized via CSS. In order to compute unsized-images failures, we need to collect
+   * matched CSS rules to see if this is the case.
+   * @url http://go/dwoqq (googlers only)
+   * @param {LH.Gatherer.FRProtocolSession} session
+   * @param {string} devtoolsNodePath
+   * @param {LH.Artifacts.ImageElement} element
+   */
+  async fetchSourceRules(session, devtoolsNodePath, element) {
+    try {
+      const {nodeId} = await session.sendCommand('DOM.pushNodeByPathToFrontend', {
+        path: devtoolsNodePath,
+      });
+      if (!nodeId) return;
+
+      const matchedRules = await session.sendCommand('CSS.getMatchedStylesForNode', {
+        nodeId: nodeId,
+      });
+      const width = getEffectiveSizingRule(matchedRules, 'width');
+      const height = getEffectiveSizingRule(matchedRules, 'height');
+      const aspectRatio = getEffectiveSizingRule(matchedRules, 'aspect-ratio');
+      element.cssEffectiveRules = {width, height, aspectRatio};
+    } catch (err) {
+      if (/No node.*found/.test(err.message)) return;
+      throw err;
+    }
+  }
+
+  /**
+   *
+   * @param {LH.Gatherer.FRTransitionalDriver} driver
+   * @param {LH.Artifacts.ImageElement[]} elements
+   * @param {Record<string, LH.Artifacts.NetworkRequest>=} indexedNetworkRecords
+   */
+  async collectExtraDetails(driver, elements, indexedNetworkRecords) {
+    // Don't do more than 5s of this expensive devtools protocol work. See #11289
+    let reachedGatheringBudget = false;
+    setTimeout(_ => (reachedGatheringBudget = true), 5000);
+    let skippedCount = 0;
+
+    for (const element of elements) {
+      // Pull some of our information directly off the network record.
+      const networkRecord = indexedNetworkRecords && indexedNetworkRecords[element.src];
+      element.mimeType = networkRecord && networkRecord.mimeType;
+
+      if (reachedGatheringBudget) {
+        skippedCount++;
+        continue;
+      }
+
+      // Need source rules to determine if sized via CSS (for unsized-images).
+      if (!element.isInShadowDOM && !element.isCss) {
+        await this.fetchSourceRules(driver.defaultSession, element.node.devtoolsNodePath, element);
+      }
+      // Images within `picture` behave strangely and natural size information isn't accurate,
+      // CSS images have no natural size information at all. Try to get the actual size if we can.
+      // Additional fetch is expensive; don't bother if we don't have a networkRecord for the image.
+      if (element.isPicture || element.isCss || element.srcset) {
+        if (indexedNetworkRecords && !networkRecord) continue;
+        await this.fetchElementWithSizeInformation(driver, element);
+      }
```

Comment:
This logic has been changed slightly. The image record will be fetched if we run in snapshot mode.